### PR TITLE
Add non-affine store for PXA dialect

### PR DESCRIPTION
**Pull request by shiyao00**

### CODE REVIEW

File: pmlc/dialect/pxa/ir/ops.td [hunk 1]
```diff
@@ -273,6 +273,49 @@ def PxaVectorLoadOp : PxaLoadOpBase<"vector_load"> {
   let hasFolder = 1;
 }
 
+class PxaStoreOpBase<string mnemonic, list<OpTrait> traits = []> :
+    PXA_OpWithPP<mnemonic, !listconcat(traits, [MemRefsNormalizable])> {
+
+  code extraClassDeclarationBase = [{
+    mlir::Value getValueToStore() { return value(); }
+    unsigned getValueOperandIndex() { return 0; }
```
Comment (flaub) [inline pmlc/dialect/pxa/ir/ops.td:281]: These are already generated by the ODS/tablegen, we shouldn't need them.

File: pmlc/dialect/pxa/ir/ops.td [hunk 2]
```diff
@@ -273,6 +273,49 @@ def PxaVectorLoadOp : PxaLoadOpBase<"vector_load"> {
   let hasFolder = 1;
 }
 
+class PxaStoreOpBase<string mnemonic, list<OpTrait> traits = []> :
+    PXA_OpWithPP<mnemonic, !listconcat(traits, [MemRefsNormalizable])> {
+
+  code extraClassDeclarationBase = [{
+    mlir::Value getValueToStore() { return value(); }
```
Comment (flaub) [inline pmlc/dialect/pxa/ir/ops.td:280]: I don't think we need to make a wrapper, we can just call `value()`

File: pmlc/dialect/pxa/ir/ops.td [hunk 3]
```diff
@@ -273,6 +273,49 @@ def PxaVectorLoadOp : PxaLoadOpBase<"vector_load"> {
...
+    unsigned getValueOperandIndex() { return 0; }
+    void setValueToStore(mlir::Value value) { setOperand(getValueOperandIndex(), value); }
+
+    mlir::Value getMemRef() { return memref(); }
+    unsigned getMemRefOperandIndex() { return 1; }
+    void setMemRef(mlir::Value value) { setOperand(getMemRefOperandIndex(), value); }
```
Comment (flaub) [inline pmlc/dialect/pxa/ir/ops.td:286]: ODS already generates a setter, just use `op.memref(newValue)`

**Review by jbruestle — CHANGES_REQUESTED** [review]
Can we add a reduction type (i.e. assign, addi, etc)?  Basically more like atomicrmw or pxa.reduce.  In that case, we can get plain 'store' via 'assign', but we also add atomic add, which is actually what scatter often uses (and it removes a load + add + store for the example you sent originally).  Also it makes thing parallel-safe if we lower to atomicrmw, which in fact scatter should require (right now I think it just produces incorrect results in some cases, but we don't have enough tests).

**Comment by shiyao00** [comment]
> Can we add a reduction type (i.e. assign, addi, etc)? Basically more like atomicrmw or pxa.reduce. In that case, we can get plain 'store' via 'assign', but we also add atomic add, which is actually what scatter often uses (and it removes a load + add + store for the example you sent originally). Also it makes thing parallel-safe if we lower to atomicrmw, which in fact scatter should require (right now I think it just produces incorrect results in some cases, but we don't have enough tests).

Done

File: pmlc/dialect/pxa/ir/ops.td (hunk 1 above)
Comment (shiyao00) [reply pmlc/dialect/pxa/ir/ops.td]: Removed

**Review by jbruestle — APPROVED** [review]

**Pull request merged**